### PR TITLE
A shopping cart model, controller, repository, and service tiers have…

### PR DESCRIPTION
**Pull request by oleksandr-chernetskyi**
… been implemented. Also, necessary files for the creation of new tables in the db.changelog-master.yaml file have been created.

### CODE REVIEW

File: src/main/java/springboot/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Override
+    public ShoppingCartResponseDto getShoppingCart() {
+        User authenticatedUser = userService.getAuthenticatedUser();
+        ShoppingCart shoppingCart = shoppingCartRepository
+                .getUserById(authenticatedUser.getId())
+                .orElseThrow(() -> {
+                    log.error("getShoppingCart method failed. Can't find shopping cart: {}",
+                            authenticatedUser.getId());
+                    return new NoShoppingCartException("Shopping cart not found for user");
```
Comment (MentorMA) [inline src/main/java/springboot/onlinebookstore/service/impl/ShoppingCartServiceImpl.java]: Did you try to throw instead of return?

**Review by MentorMA — CHANGES_REQUESTED** [review]

File: src/main/java/springboot/onlinebookstore/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+@Tag(name = "Shopping Cart Controller")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+    private final CartItemService cartItemService;
+
+    @PostMapping
+    @Operation(summary = "Add new item to shopping cart")
+    public CartItemResponseDto cartItemResponseDto(@RequestBody @Valid
+            CartItemRequestDto cartItemRequestDto) {
+        return shoppingCartService.save(cartItemRequestDto);
```
Comment (Rommelua) [inline src/main/java/springboot/onlinebookstore/controller/ShoppingCartController.java]:
```suggestion
        return shoppingCartService.addCartItem(cartItemRequestDto);
```

File: src/main/java/springboot/onlinebookstore/controller/ShoppingCartController.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    @GetMapping
+    @Operation(summary = "Get shopping cart")
+    public ShoppingCartResponseDto shoppingCartResponseDto() {
+        return shoppingCartService.getShoppingCart();
+    }
+
+    @PutMapping("/cart-items/{id}")
+    @Operation(summary = "Update cart item by ID")
+    public CartItemResponseDto update(@RequestBody @Valid
```
Comment (Rommelua) [inline src/main/java/springboot/onlinebookstore/controller/ShoppingCartController.java]:
```suggestion
    public CartItemResponseDto updateCartItem(@RequestBody @Valid
```

File: src/main/java/springboot/onlinebookstore/controller/ShoppingCartController.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+@RequiredArgsConstructor
+@RequestMapping("/cart")
+@Tag(name = "Shopping Cart Controller")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+    private final CartItemService cartItemService;
+
+    @PostMapping
+    @Operation(summary = "Add new item to shopping cart")
+    public CartItemResponseDto cartItemResponseDto(@RequestBody @Valid
```
Comment (Rommelua) [inline src/main/java/springboot/onlinebookstore/controller/ShoppingCartController.java]: Think about method name

File: src/main/java/springboot/onlinebookstore/controller/ShoppingCartController.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            CartItemRequestDto cartItemRequestDto) {
+        return shoppingCartService.save(cartItemRequestDto);
+    }
+
+    @GetMapping
+    @Operation(summary = "Get shopping cart")
+    public ShoppingCartResponseDto shoppingCartResponseDto() {
```
Comment (Rommelua) [inline src/main/java/springboot/onlinebookstore/controller/ShoppingCartController.java]: Think about method name

File: src/main/java/springboot/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import springboot.onlinebookstore.service.ShoppingCartService;
+import springboot.onlinebookstore.service.UserService;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final UserService userService;
+    private final CartItemService cartItemService;
```
Comment (Rommelua) [inline src/main/java/springboot/onlinebookstore/service/impl/ShoppingCartServiceImpl.java]: Do not use CartItemService here
```suggestion
```

File: src/main/java/springboot/onlinebookstore/service/CartItemService.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+import springboot.onlinebookstore.dto.cartitem.CountQuantityDto;
+import springboot.onlinebookstore.model.CartItem;
+import springboot.onlinebookstore.model.User;
+
+public interface CartItemService {
+    CartItemResponseDto save(CartItemRequestDto cartItemRequestDto);
+
+    Set<CartItemResponseDto> findItemsByShoppingCartId(Long id);
+
+    void setCartItemsAndShoppingCart(User user, CartItem cartItem);
```
Comment (Rommelua) [inline src/main/java/springboot/onlinebookstore/service/CartItemService.java]:
```suggestion
```

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/springboot/onlinebookstore/service/CartItemService.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package springboot.onlinebookstore.service;
+
+import java.util.Set;
+import springboot.onlinebookstore.dto.cartitem.CartItemRequestDto;
+import springboot.onlinebookstore.dto.cartitem.CartItemResponseDto;
+import springboot.onlinebookstore.dto.cartitem.CountQuantityDto;
+
+public interface CartItemService {
+    CartItemResponseDto save(CartItemRequestDto cartItemRequestDto);
```
Comment (Rommelua) [inline src/main/java/springboot/onlinebookstore/service/CartItemService.java]: I do not see usage of this method

File: src/main/java/springboot/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                .map(this::mapToCartItemResponseDto)
+                .collect(Collectors.toSet()));
+        return shoppingCartResponseDto;
+    }
+
+    private CartItemResponseDto mapToCartItemResponseDto(CartItem cartItem) {
+        CartItemResponseDto cartItemResponseDto = new CartItemResponseDto();
+        cartItemResponseDto.setId(cartItem.getId());
+        cartItemResponseDto.setBookId(cartItem.getBook().getId());
+        cartItemResponseDto.setBookTitle(cartItem.getBook().getTitle());
```
Comment (Rommelua) [inline src/main/java/springboot/onlinebookstore/service/impl/ShoppingCartServiceImpl.java]: use mapper for mapping)

**Review by Rommelua — CHANGES_REQUESTED** [review]

**Review by Vasul-Pankiv — APPROVED** [review]

**Pull request merged**